### DISCUSSION TITLE
Cannot access data in local package

### DISCUSSION THREAD

**Discussion by Slarag** [pyinstaller/pyinstaller#7853] [Help] [answered]
I have a Python application which I want to bundel with pyinstaller. `mypackage` is a local python package (located next to my main python script) which also contains some binary data and does some dynamic importing by scanning some submodules/namespaces. 

My folder structure is like this:

```
.
├── docs
│   ├── manual
├── main.py
├── README.md
├── requirements.txt
├── setup.sh
├── venv
│   └── ...
└── mypackage
    ├── package1
    ├── package2
    ├── bindata1.bin
    ├── bindata2.bin
    └── __init__.py
```

This works fine when running the python script directly, but not with the frozen version. The dynamic scanning does not work and the binaries in the module (e.g. logos for the tkinter App, etc.) are not found.

I used this command: `pyinstaller main.py --clean --noconfirm --onefile --windowed` So I tried to collect the package by appending `--collect-all mypackage`, but I get those warnings:

```
115 INFO: PyInstaller: 5.13.0
115 INFO: Python: 3.10.6
116 INFO: Platform: Linux-5.19.0-46-generic-x86_64-with-glibc2.35
116 INFO: wrote /opt/mysoftware/main.spec
117 INFO: Removing temporary files and cleaning cache in /root/.cache/pyinstaller
121 WARNING: Unable to copy metadata for mypackage: The 'mypackage' distribution was not found and is required by the application
121 WARNING: collect_data_files - skipping data collection for module 'mypackage' as it is not a package.
121 WARNING: collect_dynamic_libs - skipping library collection for module 'mypackage' as it is not a package.
142 INFO: Determining a mapping of distributions to packages...
760 WARNING: Unable to determine requirements for mypackage: The 'mypackage' distribution was not found and is required by the application
760 INFO: Extending PYTHONPATH with paths
['/opt/mysfoftware']
898 INFO: checking Analysis
898 INFO: Building Analysis because Analysis-00.toc is non existent
898 INFO: Initializing module dependency graph...
...
```

How can I get Pyinstaller to collect my package so that dynamic loading and accessing binary data works?
I also tried `--collect-submodules mypackage`, `--collect-data mypackage`, `--collect-binaries mypackage`.

I don't have those issues when using `Cx_Freeze`. There, both dynamic loading/scanning works as expected and the binaries are found. For `Cx_Freeze` I'm using the following build options:
```
build_exe_options = {
    "includes": ["mypackage"],
    "packages": ["mypackage"],
    "include_msvcr": True,
}
```

**Comment by rokm** [comment, marked as answer]
Yes, that's a limitation of how `--collect-submodules mypackage`, `--collect-data mypackage`, `--collect-binaries mypackage` work. They add calls to [`collect_submodules()`](https://pyinstaller.org/en/stable/hooks.html#PyInstaller.utils.hooks.collect_submodules), [`collect_data_files()`](https://pyinstaller.org/en/stable/hooks.html#PyInstaller.utils.hooks.collect_data_files), and [`collect_dynamic_libs()`](https://pyinstaller.org/en/stable/hooks.html#PyInstaller.utils.hooks.collect_dynamic_libs) at the top of the generated .spec file, and pass the resulting lists to the `Analysis` instance.

However, the package search paths are extended with location of your entry-point script only during instantiation of `Analysis`. Therefore, those `collect_` calls cannot find a local (non-installed) copy of `mypackage`.

There are three ways to deal with this (aside from installing `mypackage`to your venv):

1. write a hook for `mypackage`. I.e., create a directory called `pyinstaller-hooks`, and create a hook file called `hook-mypackage.py` inside it, with the following content:
```python
# pyinstaller-hooks/hook-mypackage.py
from PyInstaller.utils.hooks import collect_submodules, collect_data files
hiddenimports = collect_submodules('mypackage')
datas = collect_data_files('mypackage')
binaries = collect_dynamic_libs('mypackage')
```
and then pass the additional hooks directory to PyInstaller via [`--additional-hooks-dir ./pyinstaller-hooks`](https://pyinstaller.org/en/stable/usage.html#cmdoption-additional-hooks-dir). This is how things were done before `--colllect-*` command-line "shortcuts" were added; and it works because hooks are ran during `Analysis` instantiation, so after the package search path has been extended with location of the entry-point script.

2. Use `PYTHONPATH` environment variable to add the parent directory of `mypackage` to python's global search path, before running `PyInstaller`.

3. Edit the generated spec file and append the parent directory of `mypackage` to `sys.path` before the `collect_*` calls are made (within the context of code executed in the .spec file, there is a special variable called `SPECPATH` that points to the parent directory of the spec; you can use that as an anchor to path to parent directory of `mypackage`). Then  you build your application using the modified .spec file instead of the original .py file (which would re-generate the spec and overwrite the changes).

**Reply by Slarag** [reply]
Than you very much! I've used option 1 (the additional hook) which works fine und Windows and Linux.
However, I'm using an automated build pipeline which builds the Windows executable under Wine:

`wine cmd /c pyinstaller main.py --clean --noconfirm --onefile --windowed --additional-hooks-dir ./pyinstaller-hooks`

I can see that the hook is loaded, but it is unable to find the package:
```
26015 INFO: Loading module hook 'hook-mypackage.py' from 'Z:\\builds\\...\\mysw\\pyinstaller-hooks'...
2023 WARNING: Failed to collect submodules for 'mypackage' because importing 'mypackage' raised: FileNotFoundError: [WinError 2] File not found
```

But the folder is there when I check with `wine cmd /c dir`.

**Reply by rokm** [reply]
Is the package otherwise importable in python running under wine?

**Reply by Slarag** [reply]
Yes, I can run `wine cmd /c python -m mypackage` and the package is found.

**Reply by rokm** [reply]
Hmmm... can you try adding `on_error='raise'` argument to the `collect_submodules` call in the hook? This should raise the error, and perhaps the traceback will contain some clues as to what is going on...

**Reply by Slarag** [reply]
Ah ok, I just found the error. It is not related to PyInstaller.
It was caused by a call to `subprocess.run()`  in `mypackage.__init__.py` which could not find an external application.
